base/accumulate.jl

```diff
@@ -0,0 +1,517 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+"""
+    Accumulate(op[, v0], iter)
+
+An iterator which cumulatively applies the binary operation `op` to the iterator `iter`, in a similar manner
+to [`foldl`](@ref) but returning intermediate values.
+
+If an initial `v0` value is provided, then the first value of the iterator will be
+
+    op(v0, first(iter))
+
+Otherwise it will be
+
+    Base.reduce_first(op, first(iter))
+
+See also [`Base.reduce_first`](@ref) and [`accumulate`](@ref).
+"""
+struct Accumulate{O,V,I}
+    op::O
+    v0::V
+    iter::I
+end
+Accumulate(op, iter) = Accumulate(op, undef, iter) # use `undef` as a sentinel
```

Comment:
This is exactly what I was afraid would happen with `undef`.

Reply:
We can create another type to use as a sentinel: any suggestions for names? (I had originally used `uninitialized`, which makes slightly more sense).